runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -431,13 +428,36 @@ public static StreamingDataflowWorker fromOptions(DataflowWorkerHarnessOptions o
                       config,
                       dispatcherClient::consumeWindmillDispatcherEndpoints,
                       maxWorkItemCommitBytes));
+      computationStateCache = computationStateCacheFactory.apply(configFetcher);
+      windmillStreams =
+          windmillStreamsBuilder
+              .setProcessHeartbeatResponses(
+                  new WorkHeartbeatResponseProcessor(computationStateCache::get))
+              .build();
+      windmillServer =
+          createGrpcWindmillServerWithHealthChecks(options, windmillStreams, dispatcherClient);
     } else {
+      windmillStreams = windmillStreamsBuilder.build();
       windmillServer =
-          createWindmillServerStub(options, windmillStreamFactory, dispatcherClient, ignored -> {});
+          options.getWindmillServiceEndpoint() != null
+                  || options.getLocalWindmillHostport().startsWith("grpc:")
+              ? createGrpcWindmillServerWithHealthChecks(options, windmillStreams, dispatcherClient)
+              : new JniWindmillApplianceServer(options.getLocalWindmillHostport());
       configFetcher = new StreamingApplianceComputationConfigFetcher(windmillServer::getConfig);
+      computationStateCache = computationStateCacheFactory.apply(configFetcher);
     }
 
-    return Pair.of(configFetcher, Optional.ofNullable(windmillServer));
+    return ConfigFetcherComputationStateCacheAndWindmillClient.create(
+        configFetcher, computationStateCache, windmillServer, windmillStreams);
+  }
+
+  private static GrpcWindmillServer createGrpcWindmillServerWithHealthChecks(
+      DataflowWorkerHarnessOptions options,
+      GrpcWindmillStreams windmillStreams,
```

Comment:
can this take the builder in instead and then scheduleHealthChecks could be part of the builder instead of the object?  Would be cleaner since then it would be immutable and we don't have to think about it being called multiple times etc

Reply:
done added a build method that explicitly starts healthchecks